fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecursiveUnionQueryPlan.java

```diff
@@ -0,0 +1,549 @@
+                                  @Nonnull final EvaluationContext baseContext,
+                                  @Nonnull final Value scanTempTableReference,
+                                  @Nonnull final Value insertTempTableReference,
+                                  @Nullable final FDBRecordStoreBase<?> store,
+                                  @Nonnull Function<PTempTable, TempTable> tempTableDeserializer,
+                                  @Nullable byte[] continuationBytes) {
+            this.recursiveCursorCreator = recursiveCursorCreator;
+            this.baseContext = baseContext;
+            this.insertTempTableReference = insertTempTableReference;
+            this.scanTempTableReference = scanTempTableReference;
+            overridenEvaluationContext = withEmptyTempTable(insertTempTableReference, baseContext);
+            buffersAreFlipped = false;
+            this.store = store;
+            if (continuationBytes == null) {
+                isInitialState = true;
+                recursiveUnionTempTable = baseContext.getTempTableFactory().createTempTable();
+                overridenEvaluationContext = overrideTempTableBinding(overridenEvaluationContext, scanTempTableReference, recursiveUnionTempTable);
+                activeCursor = initialCursorCreator.apply(null, overridenEvaluationContext);
+            } else {
+                final var continuation = RecursiveUnionCursor.Continuation.from(continuationBytes, tempTableDeserializer);
+                isInitialState = continuation.isInitialState();
+                recursiveUnionTempTable = continuation.getTempTable();
+                overridenEvaluationContext = overrideTempTableBinding(overridenEvaluationContext, scanTempTableReference, recursiveUnionTempTable);
+                if (isInitialState) {
+                    activeCursor = initialCursorCreator.apply(continuation.getActiveStateContinuation().toByteString(), overridenEvaluationContext);
+                } else {
+                    if (continuation.buffersAreFlipped()) {
+                        overridenEvaluationContext = flipBuffers(baseContext, false);
+                    }
+                    activeCursor = recursiveCursorCreator.apply(continuation.getActiveStateContinuation().toByteString(), overridenEvaluationContext);
+                }
+            }
+        }
+
+        @Override
+        public void notifyCursorIsExhausted() {
+            if (isInitialState) {
+                isInitialState = false;
+            }
+            overridenEvaluationContext = flipBuffers(overridenEvaluationContext, true);
+            activeCursor = recursiveCursorCreator.apply(null, overridenEvaluationContext); // restart
+        }
+
+        @Override
+        public boolean canTransitionToNewStep() {
+            if (isInitialState) {
+                return true;
```

Comment:
How would this ever happen, if `notifyCursorIsExhausted` unsets it?

Reply:
Removed the `if` branch.